eliminate md5 usage, block new usage

BenTheElder opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind cleanup

#### What this PR does / why we need it:

- blocks new md5 usage with a `forbidigo` lint rule
- migrates most non-cryptographic usage to FNV-1a
- migrates storage test that needs to compute the hash inside a pod to sha256

#### Which issue(s) this PR is related to:
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/<kep-issue-number>

If there is no associated issue, then write "N/A".
-->

Fixes: https://github.com/kubernetes/kubernetes/issues/129652

#### Special notes for your reviewer:

Gemini assisted with the endpoints unit test, which I've reviewed and run locally, I verified it passes with the controller change and fails if I revert the controller change. All other changes were hand-authored.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Eliminate use of md5 and prevent future use of md5 in favor of more appropriate hashing algorithms.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```
/sig architecture cluster-lifecycle security
/area code-organization

BenTheElder committed:
forbid new use of md5

BenTheElder committed:
migrate kubeadm static pod hashing to fnv-1a

128bit fnv-1a should be drop-in for md5 while being fast and non-obsolete
for non-cryptographic purposes.

BenTheElder committed:
mark remaining md5 usage TODO and exclude from lint

BenTheElder commented:
/cc pacoxu HirazawaUi

HirazawaUi reviewed (commented):
The changes to kubeadm are very straightforward. This function just uses MD5 for simple file comparisons and won't affect functionality.

lgtm for kubeadm.
/lgtm

BenTheElder commented:
[fixed verify spelling error in the linter config]

> Once this PR has been reviewed and has the lgtm label, please assign [thockin](https://github.com/thockin) for approval. For more information see [the Code Review Process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process).

/assign thockin

BenTheElder committed:
switch storage e2e tests from md5 to sha256

BenTheElder committed:
switch volumemanager reconciler tests to fnv-1a

128 bit fnv-1a is a pretty good non-cryptographic drop-in for md5, it should be faster, and this test code is not sensitive anyhow

BenTheElder committed:
switch kubelet static pod internal UID to fnv

fnv-1a in 128bit should be a drop-in replacement for md5 in non-cryptographic use that is faster and not obsolete

BenTheElder commented on test/e2e/storage/testsuites/volume_io.go:
```diff
@@ -44,13 +44,13 @@ import (
 )
 
-// MD5 hashes of the test file corresponding to each file size.
+// sha256 hashes of the test file corresponding to each file size.
```
While we used fnv for other non-cryptographic hashing purposes, here we want to be able to compute the hash in the pod, and sha256 has a commandline already available in our test image.

Sha256 should be plenty fast for this purpose anyhow.

BenTheElder commented on staging/src/k8s.io/kubectl/pkg/util/util.go:
```diff
@@ -47,6 +47,9 @@ func HashObject(obj runtime.Object, codec runtime.Codec) (string, error) {
+	// TODO: finish removing existing md5 usage
+	// https://github.com/kubernetes/kubernetes/issues/129652
+	//nolint:forbidigo
 	return fmt.Sprintf("%x", md5.Sum(data)), nil
```
This utility doesn't actually seem to be used anywhere by us:
https://cs.k8s.io/?q=%5C.HashObject&i=nope&literal=nope&files=&excludeFiles=&repos=

... however it is exported

soltysh replied:
I'm fine with deleting it. Even though it's an exported function it's small enough that anyone relying on its functionality can easily copy&paste those 4 lines. Digging through history it seems this was used for rolling updates, which initially were coupled inside kubectl, but eventually the entire logic migrated over to deployment controller.

BenTheElder replied:
done + rebased

neolit123 commented on cmd/kubeadm/app/util/staticpod/utils.go:
```diff
@@ -373,7 +373,7 @@ func ManifestFilesAreEqual(path1, path2 string) (bool, string, error) {
 	}
 
-	hasher := md5.New()
+	hasher := fnv.New128a()
```
i think this may be fine becase it will result in a forced manifest upgrade for the release where this change is present.

but we'd have to run upgrade tests on this PR, to make sure.

cc @SataQiu @pacoxu @carlory @HirazawaUi

BenTheElder replied:
this hash isn't exposed to users or stored to disk AFAICT, it's just used to do a deep comparison of these two pods by comparing their hashes?

BenTheElder replied:
sha256 would be fine, but I am trying to follow the general guideline of:
- fnv-1a for non-cryptographic hashes (fast, and signals that it's not for security)
- sha256 for secure hashes (still fast, signals that it's for security)

fnv is currently used in the project. fnv1-a in 128bit is roughly equivalent to md5 while not being outdated or ambiguously for "security"

We could also look at comparing the pods some other way.

---

AFAICT this should not force any manifest upgrade, because the pods should still compare the same. We will hash both using fnv instead of md5

neolit123 replied:
ok, seems like fnv-1a 128 bit will be quite difficult to have collusions and we never got collusions with the md5 either.

/lgtm

BenTheElder replied:
Yeah, we actually even use 32 and 64 bit fnv in other areas currently but I didn't think micro-optimizing made sense here and 128 bit ~matches md5 while being explicitly non-cryptographic and current. Initially I had sha256 but it was inconsistent with the larger PR as I worked through the remainder.

neolit123 commented:
> The changes to kubeadm are very straightforward. This function just uses MD5 for simple file comparisons and won't affect functionality.
> 
> lgtm for kubeadm. /lgtm

not entirely, we need to test it.

BenTheElder committed:
switch endpoints hashing from md5 to fnv-1a

128 bit fnv-1a is a pretty good non-cryptographic drop-in for md5, it should be faster

BenTheElder committed:
switch endpointslice hashing from md5 to fnv-1a

128 bit fnv-1a is a pretty good non-cryptographic drop-in for md5, it should be faster

BenTheElder committed:
remove deadcode HashObject from k8s.io/kubectl/pkg/util

I can't find evidence that anything is using this anymore, if anyone does depend on this it is a tiny function that depends only on public symbols and they can just place a copy somewhere else.

BenTheElder commented:
Removed the dead code from staging/src/k8s.io/kubectl/pkg/util/util.go per the thread https://github.com/kubernetes/kubernetes/pull/133511#discussion_r2300333098 and rebased.

BenTheElder committed:
avoid updating when endpoint subsets are merely sorted differently

BenTheElder commented on pkg/kubelet/config/common.go:
```diff
@@ -59,7 +59,7 @@ func generatePodName(name string, nodeName types.NodeName) string {
 func applyDefaults(logger klog.Logger, pod *api.Pod, source string, isFile bool, nodeName types.NodeName) error {
 	if len(pod.UID) == 0 {
-		hasher := md5.New()
+		hasher := fnv.New128a()
```
/cc @tallclair 
we previously discussed this part with @liggitt IIRC

BenTheElder commented on staging/src/k8s.io/kubectl/pkg/util/util.go:
```diff
@@ -41,15 +39,6 @@ func ParseRFC3339(s string, nowFn func() metav1.Time) (metav1.Time, error) {
 	return metav1.Time{Time: t}, nil
 }
 
-// HashObject returns the hash of a Object hash by a Codec
```
/cc @soltysh 

kubectl change x-ref https://github.com/kubernetes/kubernetes/issues/129652#issuecomment-3225042654

I can't find any evidence that this code is used https://github.com/kubernetes/kubernetes/issues/129652#issuecomment-3215980836

AFAICT it is stale / dead code.

danwinship commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -1008,6 +1008,52 @@ func TestSyncEndpointsItemsPreexistingIdentical(t *testing.T) {
+		}},
+	})
+
+	// Add pods in a different order than in the endpoint subset.
```
The order that you add them in isn't relevant; the question is what order they get Repacked into.

You want to show that the Endpoints object you just added is sorted incorrectly according to the current hash, so just call `RepackSubsets` on it, and confirm that this swaps the order of the 2 addresses. Then you can add the pods and confirm that this didn't cause a resync.

BenTheElder commented:
Windows unit tests have been unreliable.

integration is scheduler preemption flaking again
/skip pull-kubernetes-linter-hints
/retest

danwinship commented:
/approve

SergeyKanzhelev reviewed (approved):
/approve

kubelet changes look OK

BenTheElder commented:
@HirazawaUi @neolit123 would one of you mind approving for kubeadm? I haven't changed anything in cmd/kubeadm since your last reviews with LGTMs.

> Once this PR has been reviewed and has the lgtm label, please ask for approval from [thockin](https://github.com/thockin). For more information see [the Code Review Process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process).

/assign @thockin 
For the pkg/api endpoints change.

carlory commented:
/approve
for kubeadm

soltysh reviewed (approved):
/approve
the kubectl bits

thockin reviewed (commented):
/approve
